### ISSUE TITLE
Plugin - AWS Batch

### ISSUE THREAD

**Issue by satra** [nipype/pydra#41]

**Comment by arokem** [comment]
Would [Cloudknot](https://richford.github.io/cloudknot/) be useful?

**Comment by satra** [comment]
@arokem - can we use cloudknot for arbitrary docker tasks? not just python functions?

**Comment by arokem** [comment]
No. The current API requires a Python function. But you could easily automate wrapping arbitrary operations as a system call, I guess.

**Comment by satra** [comment]
@arokem - the wrapping/system call bit i can do, it's the arbitrary docker image support that i am more curious about. this hooks into AWS Batch configuration itself.

**Comment by arokem** [comment]
Yes: you can inherit the CK docker image from another arbitrary docker image. For example, I've used CK to inherit from an image generated with neurodocker that had Freesurfer in it, to run a function that uses nipype to run recon-all on one subject, then scaling to hundreds of subjects by mapping on a list of subject IDs.

**Comment by arokem** [comment]
Does that answer your question?

**Comment by satra** [comment]
thanks @arokem - that's sounds like a pretty good use case and set of tools :) 

do you have a pointer to that example? perhaps we can use that as a test case to make it pydra-like?

**Comment by arokem** [comment]
It's here: https://github.com/uw-biomedical-ml/mri2mri_notebooks/blob/master/ck-freesurfer-segmentations.ipynb

**Comment by satra** [comment]
@mgxd - let's take a look at this relative to your experience with Batch (after voice stuff is taken care of)

**Comment by satra** [comment]
@mgxd - is this something you can take a look at?

**Comment by mgxd** [comment]
sure - i'm currently pushing on finalizing a new fmriprep release, but can take a look once that's taken care of

**Comment by mgxd** [comment]
Ok, after a brief refresher on batch and pydra, I looked into this (and subsequently cloudknot). Some thoughts:

- What level of AWS integration are we striving for here? In the most basic implementation, I could see pydra requiring users  to create all necessary batch infrastructure (queues, environments, definitions, docker images) and just sending off `Task`s via [python scripts similar to the SLURM implementation](https://github.com/nipype/pydra/blob/a847b39dbdee5115e62efce3f365844f56a35a5a/pydra/engine/helpers.py#L402-L417).
- Movement of data may be tricky, as we'd likely have to upload to a bucket, bind the appropriate data, and then transfer back locally. This may also mess with the input checksum? One thing to consider is using a bucket consistently throughout processing to avoid transfer times.
- Workflows with large dependencies would either require a subset of specialized containers tailored for specific tasks, or one monolithic container that will be used throughout)

@satra I'd love to hear what you think, and if you have a clear picture of how these pieces can fit together nicely.

**Comment by satra** [comment]
@mgxd - thanks for taking a look. indeed the full extent can get complicated.

i was thinking of the simplest possible option to start with, which is run a workflow in a container. so a split only at the workflow level (e.g., running fmriprep on multiple participants. each workflow can pull it's own data, run in the container environment ,and then push data out as part of the workflow. so no need to share filesystems, stage/pass data or such. 

from that perspective i don't know if aws batch/beanstalk/lambda are the best options.

**Comment by arokem** [comment]
Lambda is unlikely to generally be a good option for this kind of thing,
because it has rather strict RAM/timing limitations.

**Comment by dPys** [comment]
@mgxd -- if you're still working on this, you may be able to find some useful functions [here](https://github.com/dPys/PyNets/blob/master/pynets/cli/pynets_cloud.py#L23-L344) and [here](https://github.com/dPys/PyNets/blob/master/pynets/core/cloud_utils.py).

@satra and @arokem -- I also think that this enhancement, if pursued, could be extremely powerful. It seems to me that Nipype's MultiProc and AWS's SPOT/ec2Fleet technology for Batch were almost made for each other?

**Comment by mgxd** [comment]
@dPys I haven't been working on this lately, and have my plate fairly full at the moment, so if you are interested / up for it feel free to give it a go!

**Comment by dPys** [comment]
@mgxd -- I am working with a full plate as well at the moment, and will be for some time, but when the time is right, I'd be very down to work on this with you.
